vendor/parser-jane/for-parser-standard/parser.mly

```diff
@@ -153,9 +153,31 @@ let mk_attr ~loc name payload =
   Attr.mk ~loc name payload
 
 let mkpat_with_modes ~loc ~pat ~cty ~modes =
-  match cty, modes with
-  | None, [] -> pat
-  | cty, modes -> mkpat ~loc (Ppat_constraint (pat, cty, modes))
+  match pat.ppat_desc with
+  | Ppat_constraint (pat', cty', modes') ->
+    begin match cty, cty' with
+    | Some _, None ->
+      { pat with
+        ppat_desc = Ppat_constraint (pat', cty, modes @ modes');
+        ppat_loc = make_loc loc
+      }
+    | None, _ ->
+      { pat with
+        ppat_desc = Ppat_constraint (pat', cty', modes @ modes');
+        ppat_loc = make_loc loc
+      }
+    | _ ->
+      mkpat ~loc (Ppat_constraint (pat, cty, modes))
+    end
+  | _ ->
+    begin match cty, modes with
+    | None, [] -> pat
+    | cty, modes -> mkpat ~loc (Ppat_constraint (pat, cty, modes))
+    end
```

Comment:
This code is essentially creating an invariant about what is and isn't a legal parsetree wrt constraints.  If you haven't, you should check this invariant in your branch of the compiler that also has this change, in `parsing/ast_invariants.ml`.  (And add corresponding tests - see `tests/parsing/broken_invariants.ml`)

Reply:
Good idea, added a test to the compiler feature